blog/_posts/2016-12-xx-moredots.md

```diff
@@ -0,0 +1,456 @@
+
+The short answers are:
+
+1. [Ordinary vectorized code is fast, but not as fast as a hand-written loop](http://www.johnmyleswhite.com/notebook/2013/12/22/the-relationship-between-vectorized-and-devectorized-code/)
+  (assuming loops are efficiently compiled, as in Julia)
+  because each vectorized operation generates a new temporary array and
+  executes a separate loop, leading to a lot of overhead when multiple
+  vectorized operations are combined.
+
+2. The dots allow Julia to recognize the "vectorized" nature of the
+  operations at a *syntactic* level (before e.g. the type of `x` is known),
+  and hence the loop fusion is a *syntactic guarantee*, not a
+  compiler optimization that may or may not occur for carefully written code.  They also allow the *caller* to "vectorize" *any* function, rather than relying on the function author.
+
+3. Other languages have implemented loop fusion for vectorized operations,
+  but typically for only a small set of types and operations/functions that
+  are known to the compiler.  Julia's ability to do it generically, even
+  for *user-defined* array types and functions/operators, is unusual
+  and relies in part on the syntax choices above and on its ability to efficiently
+  compile higher-order functions.
+
+## Isn't vectorized code already fast?
+
+To explore this question (also discussed
+[in this blog post](http://www.johnmyleswhite.com/notebook/2013/12/22/the-relationship-between-vectorized-and-devectorized-code/)), let's begin by rewriting the code above in a more traditional vectorized style, without
+so many dots, such as you might use in Julia 0.4 or in other languages
+(most famously Matlab, Python/Numpy, or R).   
+
+```jl
+X = f(2 * X.^2 + 6 * X.^3 - sqrt(X))
+```
+
+Of course, this assumes that the functions `sqrt` and `f` are "vectorized,"
+i.e. that they accept vector arguments `X` and compute the
+function elementwise.  This is true of `sqrt` in Julia 0.4, but it
+means that we have to rewrite our function `f` from above in a vectorized style, as
+e.g. `f(x) = 3x.^2 + 5x + 2` (changing `f` to use the elementwise operator `.^` because
+`array^scalar` is not defined).   (If we cared a lot about efficiency, we
+might instead define a special method `f(X::AbstractArray) = map(f, X)`
+or use the `@vectorize_1arg f Number` macro defined in Julia 0.4.)
+
+### Which functions are vectorized?
+
+As an aside, this example illustrates an annoyance with the vectorized style:
+you have to *decide in advance* whether a given function `f(x)`
+will also be applied elementwise to arrays, and either
+write it specially or define a corresponding elementwise method.
+
+(Even if a function *accepts* an array argument `x` — our function `f`
+accepts any `x` type, and in Matlab or R there is no distinction between
+a scalar and a 1-element array — that doesn't mean it will *work* elementwise
+for an array unless you write the function with that in mind.)
```

Comment:
The separation between this sentence's pre- and post- emdash-parenthetical-phrase parts caused my wetware parser to throw and retry a couple times. Perhaps
```julia
Even if a function *accepts* an array argument `x`, that doesn't mean it will *work* elementwise for an array (unless you write the function with that in mind). For example, our function `f`'s original definition (`f(x) = 3x^2 + 5x + 2`) accepts any `x` type, but does not work for arrays.
```
would be easier to follow reading linearly? (With the `Matlab`/`R` mention integrated somehow of course.) Best!